homeassistant/helpers/area_registry.py

```diff
@@ -48,6 +48,18 @@ def async_list_areas(self) -> Iterable[AreaEntry]:
         """Get all areas."""
         return self.areas.values()
 
+    @callback
+    def async_get_or_create(self, name: str) -> AreaEntry:
```

Comment:
This needs to be approved first by the core team. Has this been discussed?

Reply:
> Have the area registry changes been discussed with the core team? I'm not sure we'll approve those.

They have not.  How would you suggest we proceed on that specifically?  (I could separate them out into a different PR).  This is incredibly useful for those of us with systems that have their own notion of areas.  Generally speaking, this was the entire motivation for this entire overhaul.

I'll go over the other requested changes and config_flow tests.

Reply:
Please open an architecture issue.
https://github.com/home-assistant/architecture/issues

Reply:
Sounds good to remove it from this PR if possible so we aren't blocked by that review.

Reply:
I've created https://github.com/home-assistant/architecture/issues/425 and a PR specifically for that functionality.

I'll rebase this PR and then split out the area integration support and work on the comments.